README

			   SAP::Rfc XS Package


    Copyright (c) 1999 - 2006 Piers Harding.
    All rights reserved.
    This program is free software; you can redistribute it and/or modify
    it under the terms of either:

	a) the GNU General Public License as published by the Free
	Software Foundation; either version 1, or (at your option) any
	later version, or

	b) the "Artistic License" which comes with this Kit.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either
    the GNU General Public License or the Artistic License for more details.

    You should have received a copy of the Artistic License with this
    Kit, in the file named "Artistic".  If not, I'll be glad to provide one.

    You should also have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software Foundation,
    Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

    For those of you that choose to use the GNU General Public License,
    my interpretation of the GNU General Public License is that no Perl
    script falls under the terms of the GPL unless you explicitly put
    said script under the terms of the GPL yourself.  Furthermore, any
    object code linked with perl does not automatically fall under the
    terms of the GPL, provided such object code only adds definitions
    of subroutines and variables, and does not otherwise impair the
    resulting interpreter from executing any standard Perl script.  I
    consider linking in C subroutines in this manner to be the moral
    equivalent of defining subroutines in the Perl language itself.  You
    may sell such an object file as proprietary provided that you provide
    or offer to provide the Perl source, as specified by the GNU General
    Public License.  (This is merely an alternate way of specifying input
    to the program.)  You may also sell a binary produced by the dumping of
    a running Perl script that belongs to you, provided that you provide or
    offer to provide the Perl source as specified by the GPL.  (The
    fact that a Perl interpreter and your code are in the same binary file
    is, in this case, a form of mere aggregation.)  This is my interpretation
    of the GPL.  If you still have concerns or difficulties understanding
    my intent, feel free to contact me.  Of course, the Artistic License
    spells all this out for your protection, so you may prefer to use that.


    One last thing - The SAP RFCSDK is a prerequisite for this perl package to work.  The RFCSDK is proprietrary software owned by SAP AG.  For this reason I will not supply ANY component part of the RFCSDK with this perl package, and further more you must obtain it through the normal channels with SAP AG - ie. you must have a licensed SAP R/3 installation at your disposal.



After all the legal stuff -  Hello!

Welcome to the SAP::Rfc perl package.  This package is intended to facilitate RFC calls to an SAP R/3 system of release 3.1x and above.  It may work for earlier versions but it hasn't been tested.
The fundamental purpose of the production of this package, is to provide a clean object oriented interface to RFC calls from within perl.  This will hopefully have a number of effects:
(1) make it really easy to do RFC calls to SAP from perl in an object oriented fashion (Doh!)
(2) promote perl as the interface/scripting/glue language of choice for interaction with SAP R/3.
(3) make the combination of Linux, Apache, and perl the killer app for internet connectivity with SAP.
(4) Establish a small fun open source project that people are more than welcome to contribute to, if they so wish.

As of April 2003 (VERSION 1.11), SAP::Rfc has had another large chunk of functionality added.  
The ability to perform registered RFC calls where ABAP code in SAP can call out to your Perl 
code.  See the description of accept() in perldoc SAP::Rfc.


With this in mind - this package has been developed under Linux, so the installation is therefore focused on this.  This does not mean that it will not work on other UNIX like flavours - to the contrary it probably will, it just hasn't been tested.  The package has been compiled and run under Win32 - special attention needs to be paid to the options found in the Makefile.PL file.

Failing this - for Win32 people Olivier Boudry has kindly supplied me (so every win32 slave out there, should be very grateful to him) with a PPD based install package from Version 1.17 onwards - these can be found at http://www.piersharding.com/download/win32/ .

Furthermore - if you still need to build your own win32 version then Olivier has given additional instructions on doing this with the "free" Visual Studio version called VS Express 2005.
This can be found at:
http://www.piersharding.com/download/win32/Build_SAP-Rfc_with_VSE2005.pdf
Many thanks go to Olivier.
As VSE 2005 and above MS compilers require that a MANIFEST be embedded in the compiled binary, the build process is slightly modified - you must add a command line parameter:
perl Makefile.PL EMBED_MANIFEST=yes

After the fundamental RFC interface has been developed, the intention is then to extend this to add interface discovery, and a BAPI abstraction layer - but this is down the track.

SAP have kindly provided the RFCSDK for Linux on the OSS ftp service.  This can be retrieved from sapserv3, currently in the "unsupported directory" - this should change before the end of 1999 to being fully supported. See OSS note 53459 for details.

*Note: 29.11.2002 - this is old info from circa 1999 - I am currently using RedHat 8.0 and either the librfc or librfccm from R3 4.6C.

This must be installed before you continue.  I use Redhat 6.1 or Slackware 7.0 - other distributions will probably work, but take heed of the instructions provided with the rfcsdk rpm files.
Once you have the rfcsdk installed carry on!

<old>
Make sure that the LIBS and INC directive in the Makefile.PL points to the correct location of your OS's version of the rfcsdk -without this linking correctly nothing is going to work.
</old>

<new>
Thanks to Achim Grolms who supplied the following information :-

- where to get the required SAP files for the RFCSDK:

<ftp://ftp.sap.com/pub/linuxlab/contrib/SAPCAR>
Archive program needed to exratct the archives

<ftp://ftp.sap.com/pub/linuxlab/contrib/librfc_mt_so_dbg.CAR>
Dynamic libraries and headers

-This is how to extract
(Like ./SAPCAR -xvf librfc_mt_so_dbg.CAR)

-and where to put them
(libs to $RFCSDKHOME/lib and headers to $RFCSDKHOME/include)

where RFCSDK should be /usr/sap/rfcsdk as a standard
</new>

After you have unpacked The SAP::Rfc module, you should have all the files listed
in MANIFEST.

Installation

In brief, the following should work on most systems:
        perl Makefile.PL
	make
	make test  << provide all your host routing information.
	make install

Help on the build process is available by doing:
        perl Makefile.PL --help

and it says:

The whole trouble with the build process is one 
of determining what your OS is, and which RFC
library needs to be linked in, and where that 
library lives.

To help with this there are two command line
options that you can use to point me in the 
right direction:

(1) --source /usr/sap/rfcsdk
  This tells me where to look for your installed
  RFCSDK.  This must be the directory that contains
  the ./lib and ./include directories of the RFCSDK
  In the include directory must be saprfc.h etc.
  By default, this looks in /usr/sap/rfcsdk

(2) --addlibs ' -lm -ldl -lpthread '
  This tells me what additional libraries need to
  be passed in at link time - this can be anything
  that you would put in LDFLAGS.
  By default this is ' -lm -ldl -lpthread '


(3) --unicode
  This tells me that you want to compile in unicode 
  support - to do this you must have the unicode version of the RFC
  library specifically librfcum.xx or librfcu.a and the libsapu16 
  library.
  (good luck!!).

Example:

perl Makefile.PL --source /opt/rfcsdk --addlibs '-lsomethingwacky'


UNICODE
========

If you really really really need it - unicode support can be built 
BUT it must be done using the librfcum eg. the special unicode version
of the RFCSDK.
Get the latest as provided by service.sap.com/connectors => see OSS                                    
note 413708.

This has only ever been tested on Linux, and is only ever likely to work 
on systems that have iconv and are Little Endian in its current form.

To build it you need something like:                                                                        
perl Makefile.PL --source=/usr/sap/rfcsdk-640-unicode --unicode                                             
for win32: perl Makefile.PL EMBED_MANIFEST=yes BINARY_LOCATION=SAP-Rfc-1.51.tar.gz --iconv=C:/iconv --unicode

Please email directly if you have problems, and are willing to help 
with testing (especially for other platforms) => piers@cpan.org.



VERSION Info supplied by Others
================================

Mark Förster writes:

  On HP-UX (11.X) use the ANSI C compiler or gcc,
  extract the RFCSDK 4.6C 32bit e.g. to /usr/sap/rfcsdk.
  If you get error messages like:
  /usr/lib/dld.sl: Can't shl_load() a library containing Thread Local
	Storage: /usr/lib/libcl.2
  then try the following solution:
  export LD_PRELOAD=/usr/lib/libcl.2


Glenn Jackman writes:
FYI, the sysadmin sent me this info about compiling on Solaris:                                                      
  I added -lrfc and /usr/lib/libCrun.so.1 to EXTRALIBS and LDLOADLIBS.                                               
  I also used gcc as opposed to the workshop compiler.  I've had better                                              
  luck with this compiler compiling Open Source software.                                                            




Cheers - Piers Harding - R/2 R/3 BASIS, IFACE, JAPH, whatever ...

email: piers@cpan.org




Changes
Revision history for Perl extension SAPRfc.

0.01  Sat May 29 17:47:22 1999
	- original version; created by h2xs 1.19

0.02  Fri Jul 16 19:27:22 1999
   Documentation for modules added
   Examples added to SAPRfc POD
   Numerous bugs fixed in interface discover
   Structure discovery added
   Exceptions added to interface definition

0.03  Mon Jul 19 05:30:00 1999
   Change SAPInterface Exports(), Imports(), Tables methods to return 
     a list of objects insteadof names
   Change SAPRfc connect() method to accept login parameters
   Fix bug in SAPStructure for updating structure with NULL records
   Fix bug with interface discovery for default values - SPACE word
     means put space in field( doh!), literals are single quoted eg.
     'X'.

0.04  Wed Jul 21 17:00:00 1999
  Completely renamed all modules to SAP::xxx standard - sorry guys
    but it is better to bite this one now than later!
  Fixed bug in structure discovery for interface discovery - it was not 
    always picking that a table had a structure.
  Fixed nasty bug in returning import parameter values.
  Created sample IDOC posting program for Purchase Orders

0.05  Sat Sep 04 10:00:00 1999
  Thanks to Frederic Huet who highlighted some bugs:
  Fixed - bad C code for import parameters
  Fixed discovery problem for Date and Time data types - these get a basic
    default value of zeros.

0.10  Sat Sep 11 22:00:00 1999
  Changed the structure routine to not pack and unpack the "whole" value
  all the time - now it keeps track of when component or overall value 
  changes in order to decide when to pack/unpack.

0.11  Sat Oct 04 19:00:00 1999
  Fixed a bug in date checking for Export parameter - did not allow 10 for
  month.
  Changed Structure.pm to default 0's for date and time data types if they are
  null.
  Cleaned up Rfc.xs to remove unused functions, and declarations.

0.12 Sometime on Monday Oct 25th 1999
  Johan Schon pointed out a problem with the t/test.t script $^W = 0 not undef.
  Fixed a problem with the discovery routines where the structures for the standard RFCs
    change between releases 3.x and 4.x+ ( RFC_GET* ).
  Used #defines to remove declarations from Rfc.xs that have caused problems with 
    compilation under Win32 - Fred.
 
0.15 Tuesday 16th  Nov 1999
  DJ sent in problems with warns coming up when using -W and use strict - a lot of these
    warnings have been fixed, but some may still exist?
  Major new release with the xml based rfc message passing ( perl <-> xs ), which also
    includes dynamic memory allocation for working storage - the method xmlrfc in Rfc.pm
    replaces callrfc if you choose to use this functionality.  callrfc remains untouched.
  dumpdef() Methods added to Interface.pm, Table.pm, Export.pm, and Import.pm to allow
    easy printing out of the object definitions - this is essentially a debugging tool.
 
 
0.16 Monday 22th  Nov 1999
  Sorry! - an annoying change has been made so that a $rfc->callrfc or a $rfc->xmlrfc now requires an arguement of an SAP::Interface object instead of the name of the interface.  This change is necessary to allow more than one instance of the same RFC call to be made within a threaded perl application.  This will eventually make the $rfc->addInterface, and $rfc->deleteInterface methods redundant, as the Interface object becomes divorced from the connectin object ( SAP::Rfc ).
  Introduced a new method in SAP::Interface to generate a code stub for a given interface - the method is $iface->Gencode($file);  It take a target source code file as an arguement.  The file is filled with a single method that returns a complete SAP::Interface object for the specified RFC Interface.  This is a time saver for complex interface definitions where you can't afford to have the overhead of using the $rfc->discover routine.
  The same method has been added to the SAP::Structure object so that structure definition code stubs can be automatically generated.  This is particularly useful for IDOC related programs that often will use many different and complex structures.  The examples program idoc.pl in the examples directory of the distribution generates a complete program stub for a given IDOC ( well worth a look ! ).

0.91 Tuesday 20th March, 2001

Completely rewritten the interface to use Inline::C


0.93 Tuesday 29th May, 2001

Added in a module SAP::SOAP that uses SOAP::Lite to support SOAP requests
Fixed INT1 and INT2 support in Structures


0.94 Wednesday 30th May, 2001

Fixed a sort order problem in SAP::Struc using cmp instead of <=> - Thank you Ivo Clarysse.
Tidied up the documentation a bit as it is way out of date.

0.96 Fri Dec 28 13:35:17 GMT 2001 

Fixed a problem with sub nextRow in SAP::Iface, which was due to earlier changes in SAP::Struc.

0.97 Fri Feb 22 09:19:00 GMT 2002 

Fixed a problem with the installer - the version nos were wrong
Added in the SECUDE login parameters

0.98 Fri Tue 26 09:00:00 GMT 2002 

Quote login paramters so that parameters with spaces in them can be passed.

0.99 Fri 19.04.2002
As per a bug fix from Jan Hebler - change the test for the return code for the Inline portion of the rfc call to a string compare - gets rid of yucky warning messages.
Fix problesm with running Inline based programs under Apache/mod_perl
Add in the ability to pass in table values ( rows ) as an array ref of hash refs - the hash refs contain a hash of the table structure key/value pairs.  The same has been done for parameters that have structures.
Note: the whole record/complex parameter values can still be passed in as strings.

1.00 Fri 05.11.2002
Remove dependency on Inline::C ( go back to using pure XS ), and control the exit on bad connects better

1.02 Fri Nov 29 12:27:45 GMT 2002
Rearranged the test suite, and fixed up makesoap.pl etc.
fixed a bug to do with export parameter structures.

1.03 Tue Dec 10 06:34:42 GMT 2002

Makefile.PL
use ExtUtils::MakeMaker;

# FIND POSSIBLE RFC INSTALLATIONS
my @POSSIBLE_RFC_DIRS = qw(/usr/sap/rfcsdk /usr/local /usr);
my $EMBED_MANIFEST= 0;
map {my($key,$val) = split(/=/,$_); $EMBED_MANIFEST = 1 if $key eq "EMBED_MANIFEST" && $val =~ /yes/ } @ARGV;

# the ADDITIONAL_LIBS needs to be modified for each OS flavour - the rfcsdk
# has the information regarding where libs and incs resides for each
# system type

#This example information comes from the readme in the rfcsdk
#HP-UX : 
#=======
#cc -Ae -I .../rfcsdk/include -L .../rfcsdk/lib
#   sapinfo.c -o sapinfo -lrfc -lcl -ldld -lM
#
#AIX (RS/6000):
#==============
#cc -I .../rfcsdk/include -L .../rfcsdk/lib
#   sapinfo.c -o sapinfo -lrfc
#
#SINIX (RM600): 
#==============
#/opt/C/bin/cc -I .../rfcsdk/include -L .../rfcsdk/lib
#   sapinfo.c -o sapinfo -lrfc -lsocket -lnsl -lusc -ldl -lm
#
#DEC AlphaOSF:
#=============
#cc -std1 -unsigned -I .../rfcsdk/include -L .../rfcsdk/lib
#  sapinfo.c -o sapinfo -lrfc -lots -lm
#
#SUN (SunPro) :
#==============
#/opt/SUNWspro/bin/cc -Xc -xcg92 -I .../rfcsdk/include -L .../rfcsdk/lib
#    sapinfo.c -o sapinfo -lrfc -lsocket -lnsl -ldl -lm -lw

# This SHOULD work for most cases (well at least Win32, and Linux)
my $ADDITIONAL_LIBS = '-lm -ldl -lpthread -lc ';

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
#  inorder for the linktype dynamic to work, perl must be built with
#  threads, and the rfc library included - see running Configure for perl

use File::Basename qw(dirname basename);
use Getopt::Long;
use vars qw($opt_source $opt_addlibs $opt_help);

&GetOptions("source=s", \$opt_source, "addlibs=s", \$opt_addlibs, "iconv=s", \$opt_iconv, "unicode", \$opt_unicode, "help", \$opt_help);

$ADD_FLAGS = "";
if ($opt_unicode){
  print "O-ohhh -unicode!!\n";
	$ADDITIONAL_LIBS .= '-lsapu16 -lstdc++';
	$ADD_FLAGS = " -DSAP_RFC_TIME -DSAPwithUNICODE -DSAPonLIN -DSAPonUNIX ";
}

my ($iconv_inc, $iconv_lib);
if ($opt_iconv){
	$iconv_inc = $opt_iconv."/include";
	$iconv_lib = $opt_iconv."/lib";
	die "iconv include dir does not exist ($iconv_inc)\n" unless -d $iconv_inc;
	die "iconv lib dir does not exist ($iconv_lib)\n" unless -d $iconv_lib;
	$iconv_inc = "-I".$iconv_inc;
	$iconv_lib = "-L".$iconv_lib." -liconv";
  print "Iconv directory specified: adding $iconv_inc and $iconv_lib\n";
}

# spew forth the help
if ($opt_help){
print <<HELP;

==================================================================
BUILDING HELP
==================================================================

Hello - you obviously need Help! :-)

The whole trouble with the build process is one 
of determining what your OS is, and which RFC
library needs to be linked in, and where that 
library lives.

To help with this there are two command line
options that you can use to point me in the 
right direction:

(1) --source /usr/sap/rfcsdk
  This tells me where to look for your installed
  RFCSDK.  This must be the directory that contains
  the ./lib and ./include directories of the RFCSDK
  In the include directory must be saprfc.h etc.
  By default, this looks in /usr/sap/rfcsdk.
  librfccm is taken in preference to librfc.

(2) --addlibs ' -lm -ldl -lpthread '
  This tells me what additional libraries need to
  be passed in at link time - this can be anything
  that you would put in LDFLAGS.
  By default this is ' -lm -ldl -lpthread '

(3) --unicode
  This tells me that you want to compile in unicode 
  support - to do this you must have the unicode version of the RFC
  library specifically librfcum.xx or librfcu.a and the libsapu16 
  library.
  (good luck!!).

(3) --iconv /some/path/to/iconv
  This says where the path is to iconv (minus the ./include|./lib)
  This path will be added to the build plus the explicit linking 
  of -liconv.

Example:

perl Makefile.PL --source /opt/rfcsdk --addlibs '-lsomethingwacky'

==================================================================

HELP
;
  exit;
}



$ADDITIONAL_LIBS = $opt_addlibs if $opt_addlibs;
unshift(@POSSIBLE_RFC_DIRS, $opt_source) if $opt_source;

my $currdir = dirname($0) || '.';
chdir($currdir) || die("can't change to $currdir: $!");
if($^O eq 'MSWin32') {
    unshift(@POSSIBLE_RFC_DIRS, 'C:/Program Files/SAP/FrontEnd/SAPgui/rfcsdk', 'C:\Program Files\SAP\FrontEnd\SAPgui\rfcsdku\lib', 'c:/rfcsdk');
}

my @CANDIDATE;
my $dir;
for $dir (@POSSIBLE_RFC_DIRS) {
    next unless -d $dir;
    my $candidate = Candidate($dir);
    next unless $candidate && $candidate->[2];
    push(@CANDIDATE, $candidate);
}

if (@CANDIDATE == 1) {
    my($dir, $inc_dir, $lib) = @{$CANDIDATE[0]};
    print "Found $lib installed in $dir\n";
} elsif (@CANDIDATE > 1) {
    if ($opt_unicode){
      print "Found the following librfcu (or librfcum) installations:\n";
		} else {
      print "Found the following librfc (or librfccm) installations:\n";
		}
    for (@CANDIDATE) {
    my($dir, $inc_dir, $lib) = @$_;
      print "\t$lib\t$dir\n";
    }
}

my $RFC_DIR;
if($opt_source && (@CANDIDATE == 1) && $CANDIDATE[0][0]) {
    $RFC_DIR = $CANDIDATE[0][0];
    print "Using --source librfc candidate found at $RFC_DIR\n";
} else {
    if($ARGV[0] =~ /^[a-z]\:|^\//i) {
    $RFC_DIR = shift;
    } else {
    unless($CANDIDATE[0][0]) {
          print "No librfc installation found, usually in $POSSIBLE_RFC_DIRS[0]\n";
    }
    $RFC_DIR = prompt "Which librfc build path do you want to link against?", $CANDIDATE[0][0];
    }
  }

my $candidate = &Candidate($RFC_DIR);
unless($candidate) {
    warn "Apparently no librfc installation at '$RFC_DIR'\nAre you sure you got it correct????\n"
      unless -f "$RFC_DIR/include/saprfc.h";
}
$candidate ||= ["$RFC_DIR", "$RFC_DIR/include", 'rfc'];
  
$RFC_INC = "$RFC_DIR/include";
  
if($^O =~ /Win32/) {
    unless($RFC_DIR =~ /^[a-zA-Z]\:/) {
    # default drive c
    $RFC_DIR = "c:".$RFC_DIR;
    }
    unless($RFC_INC =~ /^[a-zA-Z]\:/) {
      # default drive c
      $RFC_INC = "c:".$RFC_INC;
    }
    $RFC_INC = "\"$RFC_INC\"";
    $RFC_DIR =~ s|/|\\|g;
    $RFC_INC =~ s|/|\\|g;

    # patch from Ben Laurie
    if(-d "$RFC_DIR/lib") {
      $RFC_LIB = "-L\"$RFC_DIR/lib\"";
    } elsif(-d "$RFC_DIR/out32dll") {
      $RFC_LIB = "-L\"$RFC_DIR\\out32dll\"";
    }else {
    # Allow developers to point at librfc source...
      $RFC_LIB = "-L\"$RFC_DIR\"";
    }

    $SEP = "\\";
    $LIBS = ' -l'.$candidate->[2];
} else {
    $RFC_LIB = "-L".$RFC_DIR."/lib";
    $SEP = "/";
    $LIBS = ' -l'.$candidate->[2];
  
    # ccc on alpha support
    if ($^O eq 'linux' && `uname -m` =~ /alpha/ &&
        !(system("nm $RFC_DIR/lib/librfc.a|grep -q 'U _Ots'")>>8)) {
      $LIBS .= ' -lots'; 
    }
}

my $candidate_info = join('; ', @$candidate);

if($^O eq 'MSWin32' && $opt_unicode){
  $LIBS .= ' -lsapucum ';
}
  
print <<INFO;

================================================
BUILD INFORMATION
================================================

rfc dir:        $RFC_DIR
libraries:      $ADDITIONAL_LIBS $RFC_LIB $LIBS
include dir:    $RFC_INC
rfc header:     $RFC_INC/saprfc.h
rfc candidate:  $candidate_info

================================================

INFO
;


WriteMakefile(
    clean => { 'FILES'    => 'CPICTRC* ./.rfc_cache/* ./examples/.rfc_cache/* ./examples/*.trc ./examples/CPICTRC* *.inl *.trc testconn ./examples/testconn *~ core test*.txt *.tar.gz _Inline ./examples/_Inline' },
    dist => { 'TARFLAGS'    => 'cvf',
              'COMPRESS'        => 'gzip -9f',
            'SUFFIX'    => '.tgz'},
    'AUTHOR' => 'Piers Harding <piers@cpan.org>',
    'ABSTRACT' => 'Perl extension for performing RFC Function calls agains an SAP R/3 System',
    'LINKTYPE'    => 'dynamic',
    'VERSION_FROM'    => 'Rfc.pm',
    'NAME'    => 'SAP::Rfc',
#     DEFINE => ' -D'.(($^O =~ /aix/i)? 'DOIBMWKRND' : 'SUMITELSE')." $ADD_FLAGS ",
     DEFINE => '-g -Wall -D'.(($^O =~ /aix/i)? 'DOIBMWKRND' : 'SUMITELSE')." $ADD_FLAGS ",

    'LIBS'      => ["$ADDITIONAL_LIBS $RFC_LIB $LIBS $iconv_lib"],
    'INC'    => "-I".$RFC_INC ." ".$iconv_inc,

);

sub MY::makefile { '' }

sub MY::processPL { '' }


sub MY::c_o {
  my $self = shift; 
  my $section = $self->MM::c_o;
  if ($opt_unicode && $^O !~ /MSWin32/i){
    $section =~ s/(\$\*\.i)/$1i\n\t\$\(FULLPERL\) u16lit\.pl \-le \$\(BASEEXT\)\.ii/s;
		#.c$(OBJ_EXT):
		#        $(CCCMD) $(CCCDLFLAGS) "-I$(PERL_INC)" $(PASTHRU_DEFINE) $(DEFINE) $*.c
		#
    $section =~ s/(\.c\$\(OBJ_EXT\)\:.*? \$\*\.)c/$1i/s;
    #print STDERR "c_o: $section\n";
	}
  return $section;
}


# add the manifest for win32
sub MY::dynamic_lib { 
  my $self = shift; 
  my $section = $self->MM::dynamic_lib;
	if ($^O =~ /MSWin32/i && $EMBED_MANIFEST){
	  $section .= "\tcd blib\\arch\\auto\\SAP\\Rfc\n";
	  $section .= "\tmt.exe -outputresource:Rfc.dll;2 -manifest Rfc.dll.manifest\n";
	  $section .= "\tcd ..\\..\\..\\..\\..\n";
	}
  if ($opt_unicode && $^O !~ /MSWin32/i){
	  $section =~ s/(\$\(INST_DYNAMIC\)\: )/$1\$\(BASEEXT\)\.i /s;
	#  print STDERR "dynamic_lib: $section\n";
	}
	return $section;
}


# Get input values for testing
unless ($ENV{ASHOST}){
print "\n\n\n";
print "Welcome to The SAP::Rfc Object oriented Programming Interface for \n";
print "performing RFC calls to an SAP R\/3 System\.\n";
print "YOU MUST READ THE README FILE THAT COMES WITH THIS DISTRIBUTION\n";
print "AND UNDERSTAND THE TERMS AND CONDITIONS THEREIN.\n\n\n";
print "           ---------------- Please Press Enter --------------- \n";
$x=<>;
print 
    <<ENDOFTEXT;
    Copyright (c) 2002 - 2006 Piers Harding.
    All rights reserved.

    This program is free software; you can redistribute it and/or modify
    it under the terms of either:

    a) the GNU General Public License as published by the Free
    Software Foundation; either version 1, or (at your option) any
    later version, or

    b) the "Artistic License" which comes with this Kit.

    This software is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either
    the GNU General Public License or the Artistic License for more details.

    You should have received a copy of the Artistic License with this
    Kit, in the file named "Artistic".  If not, I'll be glad to provide one.

    You should also have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software Foundation,
    Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

ENDOFTEXT

print "           ---------------- Please Press Enter --------------- \n";
$x=<>;
}

print "\nPlease Enter The Following:\n";
open(FIL,">testconn") or die "cant open parm file: $!";
open(FILCP,">examples/testconn") or die "cant open parm file: $!";
&get("ASHOST", "Test SAP System Hostname:",($ENV{ASHOST} ||"localhost"));
&get("SYSNR", "System Number:",($ENV{SYSNR} || "00"));
&get("CLIENT", "Client Number:",($ENV{CLIENT} || "000"));
&get("LANG", "Login Language:",($ENV{SAPLANG} || "en"));
&get("USER", "User Id:",($ENV{SAPUSER} || "DEVELOPER"));
&get("PASSWD", "Password:",($ENV{PASSWD} || "developer"));
&get("TRACE", "Trace Level:","1");
close FIL;
close FILCP;

print "Were on our way .............\n";

# routine for each parameter
sub get{
 my $parm = shift;
 my $parmdesc = shift;
 my $parmdefault = shift;
 print "Enter $parm - $parmdesc \[$parmdefault\] ";
 my $val = <> unless $ENV{ASHOST};
 chomp $val;
 $val =~ s/^\s*?(\S+)\s*?$/$1/;
 if ( $val !~ m/^$/ ){
   print FIL "$parm\t$val\n";
   print FILCP "$parm\t$val\n";
 } else {
   print FIL "$parm\t$parmdefault\n";
   print FILCP "$parm\t$parmdefault\n";
 };
 print "\n" if $ENV{ASHOST};

}


sub Candidate {
    my $dir = shift;
    my $inc_dir;
    my @libs = ();

    for (
     "$dir/include/saprfc.h"
    ) 
      {
      if(-e $_) {
          $version_file = $_;
      } 
      }
    return unless defined $version_file;

    #$open_rfc = ($version_file =~ /openrfc/) ? 1 : 0;
    my $dirname = $inc_dir = dirname($version_file);
    return unless (-e "$dirname/saprfc.h");

    # Silly test to look for the library files
    my $libd = (-d "$dir/out32dll") ? 'out32dll' : 'lib';
    if (opendir(LIBDIR, "$dir/$libd")) {
    while (defined($_ = readdir(LIBDIR))) {
		    if ($opt_unicode){
          if($^O eq 'MSWin32') {
            push(@libs, 'rfc32u') if /rfc32u/i;
            push(@libs, 'rfcum') if /^librfcum/;
            push(@libs, 'rfcu') if /^librfcu\./;
					} else {
            push(@libs, 'rfcum') if /^librfcum/;
            push(@libs, 'rfcu') if /^librfcu\./;
					}
				} else {
          push(@libs, 'rfc32') if /rfc32/i;
          push(@libs, 'rfccm') if /^librfccm/;
          push(@libs, 'rfc') if /^librfc/;
				}
    }
    closedir(LIBDIR);
    }
		if ($opt_unicode){
      warn "$dir/lib does not seem to contain the librfcu (or librfcum) library files\n"
        unless scalar @libs;
		} else {
      warn "$dir/lib does not seem to contain the librfc (or librfccm) library files\n"
        unless scalar @libs;
		}
    @libs = reverse sort @libs;

    [$dir, $inc_dir, shift @libs];
}

